src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -13,25 +13,25 @@ namespace System.IO.Pipelines
     internal struct PipeCompletion
     {
         private static readonly ArrayPool<PipeCompletionCallback> s_completionCallbackPool = ArrayPool<PipeCompletionCallback>.Shared;
-        private static readonly Exception s_completedNoException = new Exception();
 
         private const int InitialCallbacksSize = 1;
 
-        private Exception _exception;
+        private bool _isCompleted;
+        private ExceptionDispatchInfo _exception;
 
         private PipeCompletionCallback[] _callbacks;
         private int _callbackCount;
 
-        public bool IsCompleted => _exception != null;
+        public bool IsCompleted => _isCompleted;
 
-        public bool IsFaulted => IsCompleted && _exception != s_completedNoException;
+        public bool IsFaulted => IsCompleted && _exception != null;
```

Comment:
Nit: The IsCompleted check now redundant since _exception can't be non-null if not completed.